File: smoke_tests/swift/open_redirect.swift
// Open Redirect vulnerabilities in Swift
import Foundation
import UIKit

class OpenRedirectVulnerabilities {

    // Test 1: Direct URL redirect
    func redirectToUrl(urlString: String) {
        // VULNERABLE: User-controlled redirect
        if let url = URL(string: urlString) {
            UIApplication.shared.open(url)
        }
    }

    // Test 2: Deep link handling
    func handleDeepLink(url: URL) {
        // VULNERABLE: No validation of redirect target
        if let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
           let redirectUrl = components.queryItems?.first(where: { $0.name == "redirect" })?.value,
           let targetUrl = URL(string: redirectUrl) {
            UIApplication.shared.open(targetUrl)
        }
    }

    // Test 3: OAuth callback redirect
    func handleOAuthCallback(redirectUri: String) {
        // VULNERABLE: Redirect URI from request
        if let url = URL(string: redirectUri) {
            UIApplication.shared.open(url)
        }
    }

    // Test 4: Login redirect
    func loginAndRedirect(username: String, password: String, returnUrl: String) {
        if authenticate(username: username, password: password) {
            // VULNERABLE: returnUrl from user
            if let url = URL(string: returnUrl) {
                UIApplication.shared.open(url)
            }
        }
    }

    // Test 5: Partial URL validation
    func redirectIfSafe(urlString: String) {
        // VULNERABLE: Weak validation (example.com.evil.com would pass)
        if urlString.contains("example.com") {
            if let url = URL(string: urlString) {
                UIApplication.shared.open(url)
            }
        }
    }

    // Test 6: Protocol-relative redirect
    func redirectProtocolRelative(path: String) {
        // VULNERABLE: Protocol-relative URL
        let urlString = "//\(path)"
        if let url = URL(string: "https:\(urlString)") {
            UIApplication.shared.open(url)
        }
    }

    // Test 7: WebView redirect
    func loadWebViewUrl(webView: WKWebView, urlString: String) {
        // VULNERABLE: Loading arbitrary URL
        if let url = URL(string: urlString) {
            webView.load(URLRequest(url: url))
        }
    }

    // Test 8: Logout redirect
    func logout(redirectUrl: String) {
        clearSession()
        // VULNERABLE: Post-logout redirect
        if let url = URL(string: redirectUrl) {
            UIApplication.shared.open(url)
        }
    }

    // Test 9: Share URL redirect
    func shareAndRedirect(content: String, redirectAfter: String) {
        // Share content...
        // VULNERABLE: Redirect after share
        if let url = URL(string: redirectAfter) {
            UIApplication.shared.open(url)
        }
    }

    // Test 10: Payment callback
    func handlePaymentCallback(params: [String: String]) {
        if params["status"] == "success" {
            // VULNERABLE: Redirect from payment params
            if let redirectUrl = params["redirect"],
               let url = URL(string: redirectUrl) {
                UIApplication.shared.open(url)
            }
        }
    }

    // Test 11: Error page redirect
    func showError(message: String, returnUrl: String) {
        print("Error: \(message)")
        // VULNERABLE: Return URL from parameter
        DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
            if let url = URL(string: returnUrl) {
                UIApplication.shared.open(url)
            }
        }
    }

    // Test 12: Shortlink expansion
    func expandAndRedirect(shortUrl: String) {
        // VULNERABLE: Following arbitrary shortlinks
        if let url = URL(string: shortUrl) {
            let task = URLSession.shared.dataTask(with: url) { _, response, _ in
                if let httpResponse = response as? HTTPURLResponse,
                   let location = httpResponse.allHeaderFields["Location"] as? String,
                   let redirectUrl = URL(string: location) {
                    UIApplication.shared.open(redirectUrl)
                }
            }
            task.resume()
        }
    }

    private func authenticate(username: String, password: String) -> Bool { true }
    private func clearSession() {}
}

import WebKit
class WKWebView {
    func load(_ request: URLRequest) {}
}

File: smoke_tests/lua/open_redirect.lua
-- Open Redirect vulnerabilities in Lua (OpenResty)

-- Test 1: Direct redirect from parameter
function redirect_to_url()
    local args = ngx.req.get_uri_args()
    local url = args.url
    -- VULNERABLE: Open redirect
    ngx.redirect(url)
end

-- Test 2: Redirect with next parameter
function login_redirect()
    local args = ngx.req.get_uri_args()
    local next_url = args.next or "/"
    -- Perform login...
    -- VULNERABLE: Redirect to user-controlled URL
    ngx.redirect(next_url)
end

-- Test 3: Header-based redirect
function header_redirect()
    local args = ngx.req.get_uri_args()
    local location = args.redirect_to
    -- VULNERABLE: Location header injection
    ngx.header["Location"] = location
    ngx.exit(302)
end

-- Test 4: Callback URL redirect
function oauth_callback()
    local args = ngx.req.get_uri_args()
    local callback = args.callback
    -- After OAuth processing...
    -- VULNERABLE: Callback URL from user
    ngx.redirect(callback)
end

-- Test 5: Return URL after action
function process_and_redirect()
    local args = ngx.req.get_uri_args()
    local return_url = args.return_to
    -- Process action...
    -- VULNERABLE: Return URL from user
    ngx.redirect(return_url)
end

-- Test 6: Meta refresh redirect
function meta_redirect()
    local args = ngx.req.get_uri_args()
    local target = args.target
    -- VULNERABLE: Meta refresh to user URL
    ngx.say('<meta http-equiv="refresh" content="0;url=' .. target .. '">')
end

-- Test 7: JavaScript redirect
function js_redirect()
    local args = ngx.req.get_uri_args()
    local dest = args.dest
    -- VULNERABLE: JavaScript redirect
    ngx.say('<script>window.location="' .. dest .. '";</script>')
end

-- Test 8: Partial URL validation bypass
function validated_redirect()
    local args = ngx.req.get_uri_args()
    local url = args.url
    -- VULNERABLE: Insufficient validation (can bypass with //evil.com)
    if url:sub(1, 1) == "/" then
        ngx.redirect(url)
    end
end

-- Test 9: Protocol-relative redirect
function protocol_relative()
    local args = ngx.req.get_uri_args()
    local site = args.site
    -- VULNERABLE: Protocol-relative URL
    ngx.redirect("//" .. site .. "/page")
end

-- Test 10: Redirect with referer
function referer_redirect()
    local referer = ngx.req.get_headers()["Referer"]
    -- VULNERABLE: Redirect to referer (can be spoofed)
    if referer then
        ngx.redirect(referer)
    end
end

-- Test 11: URL concatenation redirect
function concat_redirect()
    local args = ngx.req.get_uri_args()
    local path = args.path
    -- VULNERABLE: Path could be //evil.com
    local url = "https://example.com" .. path
    ngx.redirect(url)
end

-- Test 12: Form action redirect
function render_form()
    local args = ngx.req.get_uri_args()
    local action = args.action
    -- VULNERABLE: Form action from user
    ngx.say('<form action="' .. action .. '" method="POST">')
    ngx.say('<input type="submit" value="Submit">')
    ngx.say('</form>')
end

-- Test 13: Frame redirect
function frame_redirect()
    local args = ngx.req.get_uri_args()
    local frame_url = args.frame
    -- VULNERABLE: iframe src from user
    ngx.say('<iframe src="' .. frame_url .. '"></iframe>')
end
